### ISSUE TITLE
Add a lexer mode for define datas

### ISSUE THREAD

**Issue by markusamshove** [MarkusAmshove/natls#21]
Currently the parsing of datatypes is pretty hacky, especially for array dimensions as most things get recognized as identifiers.

There should be a lexer mode which makes the lexer aware when it's dealing with that stuff and omits the correct tokens instead of identifiers everywhere

- [ ] Todo category: `lexermode`

**Comment by Claes65** [comment]
Sounds a little like the "symbol table"? Can this be prioritized?

**Comment by MarkusAmshove** [comment]
I didn’t even know that it’s possible two have mutiple define data blocks. Is that a reporting mode thingy?But your natural header says structured mode Am 03.09.2024 um 16:20 schrieb Adrian Wilkins ***@***.***>:﻿
One thing I'm noticing is that I get "Unresolved reference" for all the members of anything but the first DEFINE DATA block in a file.
image.png (view on web)

**Comment by Claes65** [comment]
> I didn’t even know that it’s possible two have mutiple define data blocks. Is that a reporting mode thingy? Is that a pattern you use a lot?

It’s not possible to have 2 DD in any mode.

**Comment by awilkins** [comment]
Hmm, our code is peppered with them and I assume they catalog, but I'll check

**Comment by awilkins** [comment]
The thing they all have in common : they're all maps.

(my test was : to find all the diagnostics complaining of a "identifier" that was actually a datatype spec e.g. N8)

I wrote a test program and N1 complains about that having a second DEFINE DATA.

We have 250+ of these maps, some of them have at least 3 DEFINE DATA areas in them, all of them seem to catalog fine on a 9.3.1 UNIX container as well as the mainframe.

**Comment by MarkusAmshove** [comment]
I see, maps are also things that aren't documented well.
I wasn't able to find any documentation about INCDIC and RULEVAR.

Can you double check that other module types like subprograms can't have multiple define data's? I'll do the same with maps on our site :)

**Comment by Claes65** [comment]
Ah, processing rules. Didn’t think anyone uses those 😅 

https://documentation.softwareag.com/natural/nat827mf/edis/map_mf_proc_rules.htm

**Comment by MarkusAmshove** [comment]
The maps that we have that do contain `INCDIR` or `RULEVAR` have the following structure:

```
* >Natural Source Header 000000
* :Mode S
* :CP
* <Natural Source Header
* MAP2: PROTOTYPE              --- CREATED BY WNT 6.3.8 ---
DEFINE DATA PARAMETER
....
END-DEFINE
FORMAT PS=022 LS=080 ZP=OFF SG=OFF KD=ON  IP=OFF
* MAP2: MAP PROFILES *****************************        200***********
* ,TTAAAMMOO   D I D I N D I D I        ?_)^&:+(                       *
* 024079        N0NNUCN             X        01 SYSDBA  YR             *
************************************************************************
INPUT ....
* MAP2: VALIDATION *****************************************************
RULEVAR D01VIEW.VARIABLE
* CDIR DDM                   VARIABLE
* MAP2: END OF MAP *****************************************************
END
```

My assumption therefore was:
- DEFINE DATA
- FORMAT (optional)
- INPUT
- INCDIR/RULEVAR

The parser should already parse any statements in the map, as we don't make special distinctions in this file type.

So the only difference is that you can have more `DEFINE DATA`s in the map. Do they always come *after* the `INPUT`?

Could it be that `INCDIC` indicates that a new `DEFINE DATA` is following?

**Comment by awilkins** [comment]
TLDR

- You're allowed multiple DEFINE DATA LOCAL blocks in a map
- `INCDIC;` creates a new scope
    - Variables in the top scope do not stop being available
    - Variables declared locally shadow variables with the same name from the top scope
- `INCDIC;` is only allowed as the first statement in a RULEVAR (afaict) and the semicolon is required
    - Not sure about the whitespace!
- `RULEVAR` blocks without an `INCDIC;` share the top scope
    - ie, variables declared here are also available to everything else using the top scope

---

(all of this is inferred from error markers that N1 adds to the code when you edit it)

I tried commenting out the INCDIC from the source of one of these ; N1 doesn't add any problem markers, and it still catalogs.

If you remove a RULEVAR you start to get problems when the variables in the next `DEFINE DATA LOCAL` block have the same names as ones in prior blocks. But it doesn't complain about DEFINE DATA having to be the first block like a normal program does. Even if you add multiple DEFINE DATA LOCAL blocks to a RULEVAR program, it doesn't complain.

So I'm inferring ...

- ~~each RULEVAR has it's own scope, because you can define a variable twice in one map so long as they are in separate RULEVAR blocks~~
- You can declare a DEFINE DATA LOCAL block repeatedly in ~~RULEVAR blocks~~ MAPs
    - This is true at the top of the map, after the DEFINE DATA PARAMETER block and before the FORMAT and INPUT statements, not just in RULEVAR blocks

`INCDIC` is not legal twice in a RULEVAR block, and seems to only be legal as the first statement.

If you define a variable in the top scope (e.g. in that second DEFINE DATA LOCAL block under the PARAMETER one), it collides with the ones in the lower scopes .. unless you have the INCDIC.

e.g.

If I add a loop index at the top and a RULEVAR program redeclares it, this is an error without the INCDIC. The example I'm looking at has the same loop variable name declared with different types in several of the RULEVAR, so `INCDIC   ;` seems to mean "I want my own scope" and without it you inhabit the top scope. If you declare the same variable in two blocks without an INCDIC, they collide.

Also the semicolon after the INCDIC is part of the syntax. All ours are like this, I don't know if the space is significant too.

```
RULEVAR F00L-PARAM-AREA.L-PAGE-START
INCDIC                                 ;
```
